src/Cardano/CoinSelection/LargestFirst.hs

```diff
@@ -20,77 +21,196 @@ import Cardano.Types
     ( left )
 import Control.Monad
-    ( foldM, when )
+    ( foldM )
 import Control.Monad.Trans.Except
     ( ExceptT (..), except, throwE )
 import Data.Functor
     ( ($>) )
 import Data.List.NonEmpty
     ( NonEmpty (..) )
 import Data.Ord
-    ( Down (..), comparing )
+    ( Down (..) )
 
 import qualified Data.List as L
 import qualified Data.List.NonEmpty as NE
 import qualified Data.Map.Strict as Map
 
--- | Implements the __Largest-First__ coin selection algorithm.
+-- | Generate a coin selection according to the __largest first__ algorithm.
+--
+-- === Summary
+--
+-- For the given /output list/ and /initial UTxO set/, this algorithm generates
+-- a /coin selection/ that is capable of paying for all of the outputs, and a
+-- /remaining UTxO set/ from which spent values have been removed.
+--
+-- === State Maintained by the Algorithm
+--
+-- At all stages of processing, the algorithm maintains:
+--
+--  1.  A __/remaining UTxO list/__
+--
+--      This is initially equal to the given /initial UTxO set/ parameter,
+--      sorted into /descending order of coin value/.
+--
+--      The /head/ of the list is always the remaining UTxO entry with the
+--      /greatest coin value/.
+--
+--      Entries are incrementally removed from the /head/ of the list as the
+--      algorithm proceeds, until the list is empty.
+--
+--  2.  An __/unpaid output list/__
+--
+--      This is initially equal to the given /output list/ parameter, sorted
+--      into /descending order of coin value/.
+--
+--      The /head/ of the list is always the unpaid output with the
+--      /greatest coin value/.
+--
+--      Entries are incrementally removed from the /head/ of the list as the
+--      algorithm proceeds, until the list is empty.
+--
+--  3.  An __/accumulated coin selection/__
+--
+--      This is initially /empty/.
+--
+--      Entries are incrementally added as each output is paid for, until the
+--      /unpaid output list/ is empty.
+--
+-- === Cardinality Rules
+--
+-- The algorithm requires that:
+--
+--  1.  Each output from the given /output list/ is paid for by /one or more/
+--      entries from the /initial UTxO set/.
+--
+--  2.  Each entry from the /initial UTxO set/ is used to pay for /at most one/
+--      output from the given /output list/.
+--
+--      (A single UTxO entry __cannot__ be used to pay for multiple outputs.)
+--
+-- === Order of Processing
+--
+-- The algorithm proceeds according to the following sequence of steps:
+--
+--  *   /Step 1/
+--
+--      Remove a single /unpaid output/ from the head of the
+--      /unpaid output list/.
+--
+--  *   /Step 2/
+--
+--      Repeatedly remove entries from the head of the /remaining UTxO list/
+--      until the total value of entries removed is /greater than or equal to/
+--      the value of the /unpaid output/.
+--
+--  *   /Step 3/
+--
+--      Use the /removed UTxO entries/ to pay for the /unpaid output/, by
+--      adding entries to the /accumulated coin selection/.
```

Comment:
What kind of entries are added to the accumulated coin selection?

We should make this clearer.

Reply:
Fixed in https://github.com/input-output-hk/cardano-coin-selection/pull/7/commits/49339a3dbc7c0bef65106d9a909a3a8b7d0b36ab.